### PR TITLE
Create user api

### PR DESCRIPTION
**Pull request by alphinroy-kv**

### CODE REVIEW

File: src/authentication/authentication.graphql [hunk 1]
```diff
@@ -1,6 +1,10 @@
 type Mutation {
     passwordLogin(input: UserPasswordLoginInput!): TokenResponse
     passwordSignup(input: UserPasswordSignupInput!): UserSignupResponse
+    inviteTokenSignup(input: UserInviteTokenSignupInput): InviteTokenResponse
+    refreshInviteToken(id: ID!):InviteTokenResponse
+    setPasswordForInvite(input: UserPasswordForInviteInput): UserSignupResponse
+    revokeToken(id: ID!): String
```
Comment (doylefermi-kv) [inline src/authentication/authentication.graphql:7]: Can we make this `revokeInviteToken`?

File: src/authentication/authentication.helper.ts [hunk 2]
```diff
@@ -59,4 +59,16 @@ export class AuthenticationHelper {
   generatePasswordHash(plainTextPassword: string, salt = 10) {
     return bcrypt.hashSync(plainTextPassword, salt);
   }
+
+  generateInvitationToken = (payload: any, time?: string) => {
+    return jwt.sign(payload, process.env.JWT_SECRET || '', {
```
Comment (doylefermi-kv) [inline src/authentication/authentication.helper.ts:64]: Can we use `configService` to obtain the values for env variables?

File: src/authentication/service/password.auth.service.ts [hunk 3]
```diff
@@ -1,14 +1,17 @@
...
+  UserInviteTokenSignupInput,
+  UserPasswordForInviteInput,
   UserPasswordLoginInput,
   UserPasswordSignupInput,
   UserSignupResponse,
 } from '../../schema/graphql.schema';
 import User from '../../authorization/entity/user.entity';
 import UserService from '../../authorization/service/user.service';
-import { UserNotFoundException } from '../../authorization/exception/user.exception';
+import { PasswordAlreadySetException, UserNotFoundException } from '../../authorization/exception/user.exception';
```
Comment (doylefermi-kv) [inline src/authentication/service/password.auth.service.ts:14]: Can we rerun `npm run lint` and confirm if everything is good?

File: src/authentication/service/password.auth.service.ts [hunk 4]
```diff
@@ -56,6 +59,53 @@ export default class PasswordAuthService implements Authenticatable {
     return this.userService.createUser(userFromInput);
   }
 
+  async inviteTokenSignup(
+    userDetails: UserInviteTokenSignupInput,
+  ): Promise<InviteTokenResponse> {
+    const verifyObj = await this.userService.verifyDuplicateUser(
```
Comment (doylefermi-kv) [inline src/authentication/service/password.auth.service.ts:65]: Please rename this variable `verifyObj`.

File: src/authentication/service/password.auth.service.ts [hunk 5]
```diff
@@ -56,6 +59,53 @@ export default class PasswordAuthService implements Authenticatable {
...
+    userFromInput.middleName = userDetails.middleName;
+    userFromInput.lastName = userDetails.lastName;
+    userFromInput.status = Status.INVITED;
+    const savedUser = await this.userService.createUser(userFromInput);
+    const token = this.authenticationHelper.generateInvitationToken(
+      { id: savedUser.id },
+      '7d',
+    );
+    await this.userService.updateField(savedUser.id, 'inviteToken', token);
+    return { inviteToken: token };
```
Comment (doylefermi-kv) [inline src/authentication/service/password.auth.service.ts:88]: Can we return the expiry time of the token as well?

File: src/authentication/service/password.auth.service.ts [hunk 6]
```diff
@@ -56,6 +59,53 @@ export default class PasswordAuthService implements Authenticatable {
...
+  }
+
+  async setPasswordForInvite(
```
Comment (doylefermi-kv) [inline src/authentication/service/password.auth.service.ts:91]: Can we rename this to `setPasswordForInvitedUser` ?

File: src/authentication/service/token.service.ts [hunk 7]
```diff
@@ -29,6 +29,16 @@ export class TokenService {
...
   }
 
+  async refreshInviteToken(id: string): Promise<InviteTokenResponse> {
+    const userDetails = await this.userService.getUserById(id);
+    const token = this.authenticationHelper.generateInvitationToken(
+      { id: userDetails.id },
+      '7d',
+    );
+    await this.userService.updateField(userDetails.id, 'inviteToken', token);
+    return { inviteToken: token };
```
Comment (doylefermi-kv) [inline src/authentication/service/token.service.ts:39]: Can we return the expiry time of the token as well?

File: src/authentication/service/token.service.ts [hunk 8]
```diff
@@ -39,4 +49,9 @@ export class TokenService {
     const tokenResponse: TokenResponse = { ...token, user: userRecord };
     return tokenResponse;
   }
+
+  async revokeToken( id: string ): Promise<string> {
+    await this.userService.updateField(id, 'inviteToken', null);
+    return 'Success';
```
Comment (doylefermi-kv) [inline src/authentication/service/token.service.ts:55]: Can we return a boolean value instead of a hardcoded string?

**Review by doylefermi-kv — CHANGES_REQUESTED** [review]
- Can we write tests to cover the new functionality introduced?
- Please run `npm run lint` to fix all linter warnings and code formatting.

File: src/authentication/service/token.service.ts [hunk 9]
```diff
@@ -39,4 +59,9 @@ export class TokenService {
     const tokenResponse: TokenResponse = { ...token, user: userRecord };
     return tokenResponse;
   }
+
+  async revokeInviteToken(id: string): Promise<boolean> {
+    await this.userService.updateField(id, 'inviteToken', null);
+    return true;
```
Comment (doylefermi-kv) [inline src/authentication/service/token.service.ts:65]: Can we correct the mismatch in schema and the return type?

File: src/authentication/authentication.graphql [hunk 10]
```diff
@@ -1,6 +1,10 @@
 type Mutation {
     passwordLogin(input: UserPasswordLoginInput!): TokenResponse
     passwordSignup(input: UserPasswordSignupInput!): UserSignupResponse
+    inviteTokenSignup(input: UserInviteTokenSignupInput): InviteTokenResponse
+    refreshInviteToken(id: ID!):InviteTokenResponse
+    setPasswordForInvite(input: UserPasswordForInviteInput): UserSignupResponse
+    revokeInviteToken(id: ID!): String
```
Comment (doylefermi-kv) [inline src/authentication/authentication.graphql:7]: Let's ensure it returns a mandatory value and not null in graphql. Same with the other mutations as well.

File: src/authentication/service/password.auth.service.ts [hunk 11]
```diff
@@ -56,6 +62,62 @@ export default class PasswordAuthService implements Authenticatable {
...
+    const savedUser = await this.userService.createUser(userFromInput);
+    const invitationToken = this.authenticationHelper.generateInvitationToken(
+      { id: savedUser.id },
+      '7d',
+    );
+    await this.userService.updateField(
+      savedUser.id,
+      'inviteToken',
+      invitationToken.token,
+    );
```
Comment (doylefermi-kv) [inline src/authentication/service/password.auth.service.ts:94]: Should we first create a user and then update the token or should we create a new user with the token making the operation atomic?
- How do we address the scenario when user creation without password was successful but creation and updating invite token failed.

**Review by doylefermi-kv — CHANGES_REQUESTED** [review]

File: src/authentication/resolver/user.auth.resolver.ts [hunk 12]
```diff
@@ -48,6 +53,36 @@ export default class UserAuthResolver {
...
+  @Mutation('setPasswordForInvite')
+  async setPasswordForInvite(
+    @Args('input', new ValidationPipe(UserPasswordForInviteInputSchema))
+    request: UserPasswordForInviteInput,
+  ): Promise<UserSignupResponse> {
+    return this.passwordAuthService.setPasswordForInvitedUser(request);
+  }
+
+  @Mutation('refreshInviteToken')
+  async refreshInviteToken(
```
Comment (JoyalAJohney-KeyValue) [inline src/authentication/resolver/user.auth.resolver.ts:73]: Let's keep edit-user permission for this mutation

File: src/authentication/resolver/user.auth.resolver.ts [hunk 13]
```diff
@@ -48,6 +53,36 @@ export default class UserAuthResolver {
...
+    @Args('id', ParseUUIDPipe) id: string,
+  ): Promise<InviteTokenResponse> {
+    return this.tokenService.refreshInviteToken(id);
+  }
+
+  @Mutation('revokeInviteToken')
+  async revokeInviteToken(
```
Comment (JoyalAJohney-KeyValue) [inline src/authentication/resolver/user.auth.resolver.ts:80]: Let's keep edit-user permission for this mutation

**Review by doylefermi-kv — APPROVED** [review]

**Review by JoyalAJohney-KeyValue — APPROVED** [review]

**Pull request merged**